flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GeneratedAggregations.scala

```diff
@@ -36,6 +36,15 @@ abstract class GeneratedAggregations extends Function {
   def setAggregationResults(accumulators: Row, output: Row)
 
   /**
+    * Calculates the results from accumulators, and set the results to the output (with key offset)
+    *
+    * @param accumulators the accumulators (saved in a row) which contains the current
+    *                     aggregated results
+    * @param output       output results collected in a row
+    */
+  def setAggregationResultsWithKeyOffset(accumulators: Row, output: Row)
```

Comment:
This sounds a very good idea. I actually have thought to merge *WithKeyOffset functions into the existing functions. It works for most functions, but `setAggregationResults` and `setAggregationResults` are a little tricky. For `accumulate` and `setAggregateResults`, they do not need keyOffset, but for `merge`, they need.